Add definitions of ItemData and ItemAffixData structs + one enum used for affixes

Predelnik opened the pull request:
These structs are mostly trivial but I found it a bit annoying that naming of their fields are a bit confusing in freeablo for example, so I think defining them is quite useful. Feel free to suggest a change for any name or explanation and correct my mistakes if I'm wrong.

mewmew commented:
Still trying to find good time to review this PR. Sorry for the delay, and thanks for sending the PR.

Don't want to rush it as I would prefer to keep conducting thorough PR reviews, as these structs build the corner stones for future documentation and reversing efforts.

mewmew commented:
Hi @Predelnik,

I started to take an initial look at your PR today. It looks good!

In general I think the most difficult task with organizing these notes will be to come up with good names, and to keep them consistent. To facilitate this effort, we could try to take a look at the debug information contained within the PSX release of Diablo 1 (as reported in #1), and cross reference it to the Windows release.

Note, not all structure definitions will have a one-to-one match, as the Playstation 1 had very limited memory and as such, the developers went to great lengths to reduce memory footprint. This is also the reason why the entire PSX game is not present in memory at once, but rather split into independent parts which are loaded and released when needed.

`diabpsx.bin` is the main binary, and it is always kept in memory. Then there are four libraries which are loaded and released at run-time, so that only one of these libraries is in memory at the same time. 
* `FMV.BIN`
* `FRONTEND.BIN`
* `GAME.BIN`
* `PREGAME.BIN`

Related to this PR, the `ItemData` struct seems to be present in the PSX debug info

From https://github.com/sanctuary/psx/blob/a1325bbf7a9427577db562ddfca3842d3f9504fc/types.h#L3201

```c
typedef struct ItemDataStruct {
	unsigned char iRnd;
	char iClass;
	char iLoc;
	unsigned char iCurs;
	char itype;
	char iItemId;
	unsigned short iName;
	unsigned short iSName;
	char iMinMLvl;
	unsigned char iDurability;
	unsigned char iMinDam;
	unsigned char iMaxDam;
	unsigned char iMinAC;
	unsigned char iMaxAC;
	char iMinStr;
	char iMinMag;
	char iMinDex;
	long iFlags;
	unsigned char iMiscId;
	unsigned char iSpell;
	unsigned char iUsable;
	unsigned short iValue;
	unsigned short iMaxValue;
} ItemDataStruct;
```

While I think we can infer many of the names for members from this debug info, I also think some parts have been made more clear by the new names.

For instance, comparing `drop_rate` to `iRnd`, I would prefer `drop_rate` as it is clearer, and avoids confusion with random seed values.

What do you think?

Predelnik commented:
Looks like `Class` is used instead of our `item_category` which we may use as well to avoid confusion (I obviously mean `item_class`) 

`MiscId` instead of `item_code` looks like the name itself suggest arbitrariness of this type of id, can't say if it's perferable to stick with original one or not.

Looks like something like `Equip Location` was used instead of `item_equip_type`, I'm not sure that it's better or not. Name similar to this is used in freeablo though so maybe.

`ItemId` is named in our case `unique_base_item` because it seems to be used only for associating unique items with base ones, so if this is the case I think it's better to stick with current choice.

`Cursor` instead of `item_graphics_id`, understandable because file is named `objcurs.cel` but other than that it's sort of misleading and probably named that way for historical reasons.

I like `price` is better than `Value` but `max_price` may be used for second one, I don't remember it being bigger than the first one, but it probably should be the case.

`Usable` vs `triggerable`. Probably `usable` is better, but it's sort of unobvious that money is actually usable for splitting them :)

The rest seems to be just shorter or the same versions of what I have, correct me if I missed something.

mewmew commented:
> Looks like Class is used instead of our item_category which we may use as well to avoid confusion (I obviously mean item_class)

Sure, lets use `item_class` instead of `item_category`.


> MiscId instead of item_code looks like the name itself suggest arbitrariness of this type of id, can't say if it's perferable to stick with original one or not.

I slighty prefer `item_code` compared to MiscId, mostly because it has an `item_` prefix and `item_misc_id` would be a bit long. It is true that the items specified by the `item_code` enumerate are somewhat arbitrary; basically any item which is not a weapon or armor?

> Looks like something like Equip Location was used instead of item_equip_type, I'm not sure that it's better or not. Name similar to this is used in freeablo though so maybe.

First I was going to suggest that we change to `item_equip_loc`, but thinking about it and looking at the enums, `item_equip_type` seems better, since we have `ITEM_EQUIP_TYPE_ONE_HANDED` and `ITEM_EQUIP_TYPE_TWO_HANDED`, but not `ITEM_EQUIP_LOC_LEFT_HAND` and `ITEM_EQUIP_LOC_RIGHT_HAND`. Specifying left and right hand would be equip location, and specifying the type of weapon or armor, or equipable, or `item_equip_type` in my mind. (or `item_equip_kind`, `item_equip_class`, ...).

So, I think we can keep the name as it. Thoughts?

> ItemId is named in our case unique_base_item because it seems to be used only for associating unique items with base ones, so if this is the case I think it's better to stick with current choice.

Sounds good to me. Lets stick with `unique_base_item` for now.

> Cursor instead of item_graphics_id, understandable because file is named objcurs.cel but other than that it's sort of misleading and probably named that way for historical reasons.

I agree. Objcurs seems to be a historical artifact. I think the rationale is that the mouse cursor is replaced with the item graphic when you lift an item. I definitely prefer `item_graphic_id`.

> I like price is better than Value but max_price may be used for second one, I don't remember it being bigger than the first one, but it probably should be the case.

Price sounds good to me. Lets verify if `max_price` makes sence, based on the data.

> Usable vs triggerable. Probably usable is better, but it's sort of unobvious that money is actually usable for splitting them :)

I prefer usable :)

> The rest seems to be just shorter or the same versions of what I have, correct me if I missed something.

By the looks of it the field `required_vitality` never existed, and it is probably memory padding we are seeing in the struct, as there are three 8-bit integers followed by a 32-bit integer, thus padding 8-bits in between.

Personally, I don't mind keeping required_vitality. But seeing as it is never used, we might as well get rid of it.

Predelnik commented:
> Edit, actually I am just fine with having item_misc_id as well. It's up to you :)
Yeah let's change it, I think both terms are sort of equal so let's go with technically original one.

> So, I think we can keep the name as it. Thoughts?
Yeah as far as I remember I thought that current version is better when I was revising `Item`, so I'm okay with it.

> Lets verify if max_price makes sence, based on the data.
I checked thati it's always bigger. I may have to revise my comment a bit because for rags it's actually 5 times larger than its normal price but anyway it's unused.

> I prefer usable :)
Fine.

I'll definetily remove `required_vitality`, that's just me extrapolating things :)

Okay so if you'll look through `ItemAffixData` (which seems to be missing in PSX) I will apply all the agreed changes to this PR and convert everything to cpp files etc.

mewmew commented:
> I'll definetily remove `required_vitality`, that's just me extrapolating things in weird way :)

:)

> Okay so if you'll look through `ItemAffixData` (which seems to be missing in PSX) I will apply all the agreed changes to this PR and convert everything to cpp files etc.

Sure! I'll try to take a look later this evening. I did a search for *power* which they seem to be using for item prefixes and affixes, and managed to locate this struct, which seems to be the one they used for `ItemAffixData`. Haven't had time to really take a deep look at it yet, but most of the names you've come up with seem much clearer than the original ones. We can discuss these in more detail.

```c
typedef struct PLStruct {
	int PLName;
	int PLPower;
	int PLParam1;
	int PLParam2;
	char PLMinLvl;
	long PLIType;
	unsigned char PLGOE;
	unsigned char PLDouble;
	unsigned char PLOk;
	int PLMinVal;
	int PLMaxVal;
	int PLMultVal;
} PLStruct;
```

Predelnik commented:
Oh that's really nice! Thanks for the find.
Item type flags is of type `long`, that's pretty surprising.

`excluded_combination` I actually took from freeablo so I'm not sure if it's good or not. `GOE` otoh seems a bit hard to decipher :thinking:

`double_chance` is `PLDouble` which seems pretty similar, that's good.

Not sure that `is_cursed` is `PLOk`. Seems a bit contradictory.

The rest seems to be mostly okay n the first glance. Maybe we could argue `Param` vs `Value` for min/max (but definetily not 1/2 😃  )

mewmew commented:
As the three 32-bit integer sequence `effect_type`, `min_effect_value` and `max_effect_value` is occurring in other places (e.g. the type definition for `UniqueItemData` contains an array of six effects), I would suggest we put these into a dedicated structure, perhaps `ItemEffect`?

For reference, this is the definition of UniqueItemData from the PSX release:

```c
typedef struct UItemStruct {
	int UIName;
	char UIItemId;
	char UIMinLvl;
	char UINumPL;
	int UIValue;
	char UIPower1;
	int UIParam1;
	int UIParam2;
	char UIPower2;
	int UIParam3;
	int UIParam4;
	char UIPower3;
	int UIParam5;
	int UIParam6;
	char UIPower4;
	int UIParam7;
	int UIParam8;
	char UIPower5;
	int UIParam9;
	int UIParam10;
	char UIPower6;
	int UIParam11;
	int UIParam12;
} UItemStruct;
```

And our version may look something along the lines of:

```c
typedef struct {
	const char *name;
	unique_base_item unique_base_item;
	int8_t quality_lvl;
	int8_t neffects; // number of effects.
	int8_t reserved; // padding.
	int32_T price;
	ItemEffect[6] effects;
} UniqueItemData;
```

where `ItemEffect` may be defined as:

```c
typedef struct {
typedef struct {
	item_effect_type type;
	int min_value;
	int max_value;
} ItemEffect;
```

What do you think?

mewmew commented on enumerates.md:
```diff
@@ -621,6 +621,20 @@ typedef enum {
+    ITEM_TYPE_FOR_AFFIX_WEAPON  = 0x1000  ,
+    ITEM_TYPE_FOR_AFFIX_SHIELD  = 0x10000 ,
+    ITEM_TYPE_FOR_AFFIX_ARMOR   = 0x100000,
+} item_type_for_affix;
```
Consider renaming to `affix_type` or `item_affix_type`, mostly to make the enum names and ItemAffixData struct member name a bit shorter.

```c
// Bit flag which is used to determine if affix may occur on certain item
// categories.
typedef enum {
	AFFIX_TYPE_JEWELRY      = 0x1     ,
	AFFIX_TYPE_BOW          = 0x10    ,
	AFFIX_TYPE_STAFF        = 0x100   ,
	AFFIX_TYPE_MELEE_WEAPON = 0x1000  , // sword, axe, mace
	AFFIX_TYPE_SHIELD       = 0x10000 ,
	AFFIX_TYPE_ARMOR        = 0x100000,
} affix_type;
```

Predelnik replied:
The problem with `affix_type` is that it sounds like "type of affix" rather than "type of item which is used for affixes". Not that I'm particulary fond with my naming either. Maybe `affix_item_type` then? 🤔

mewmew replied:
sure `affix_item_type` sounds good!

mewmew commented on structures.md:
```diff
@@ -184,6 +184,113 @@ References:
+    // offset 0010 (4 bytes)
+    int quality_level;
+    // offset 0014 (4 bytes)
+    item_type_for_affix possible_item_type_flags; // bitmask
```
Base on the `affix_type` comment above.

```c
affix_type type; // bitmask
```

or potentially

```c
item_affix_type type; // bitmask
```

Predelnik replied:
Fine, I'm not entirely sure if name itself should be used to explain that it's bitmask or comment is enough.

mewmew replied:
I generally think that a comment is enough to convey information about bitmasks.

mewmew replied:
Actually, thinking about it further, I believe we flags may be good.

mewmew commented on structures.md:
```diff
@@ -184,6 +184,113 @@ References:
+    // offset 0008 (4 bytes)
+    int min_effect_value;
+    // offset 000C (4 bytes)
+    int max_effect_value;
```
Consider putting `effect_type`, `min_effect_value` and `max_effect_value` into an `ItemEffect` struct.

```c
// offset 0004 (12 bytes)
ItemEffect effect;
```

Predelnik replied:
Yeah agree that would make unique item data definition clearer. The only argument against it is that it probably wasn't that way in original source code, but that's irrelevant, so I'm on board

mewmew replied:
> The only argument against it is that it probably wasn't that way in original source code, but that's irrelevant, so I'm on board

That would be my reasoning as well. Seeing as we already update names to be clearer, I'd be happy with changing it :) Glad you are too.

mewmew commented on structures.md:
```diff
@@ -184,6 +184,113 @@ References:
+    // offset 000C (4 bytes)
+    int max_effect_value;
+    // offset 0010 (4 bytes)
+    int quality_level;
```
The quality_level seems to be an 8-bit integer.

Update to

```c
// offset 0010 (1 byte)
int8_t quality_level
// offset 0011 (3 bytes)
int8_t padding_0011[3]
```

Predelnik replied:
Yeah it's true, my bad.

Predelnik commented:
So I've applied all the changes agreed on till now and reapplied changes to `.h` files.

mewmew commented on structs.h:
```diff
@@ -173,6 +173,107 @@ typedef struct { // size = 0x170
+    // offset 003C (4 bytes)
+    spell_id scroll_spell_id; // For books and staves it applied during item creation
+    // offset 0040 (4 bytes)
+    bool32_t is_triggerable; // Set if item could be triggered with right click (e.g. for gold,
```
```diff
-is_triggerable
+is_usable
```

mewmew commented on structs.h:
```diff
@@ -173,6 +173,107 @@ typedef struct { // size = 0x170
+    // offset 0000 (4 bytes)
+    item_drop_rate drop_rate;
+    // offset 0004 (1 byte)
+    item_class _class;
```
Wish we didn't have to prefix the member with underscore, but this is needed I guess since `class` is a keyword in C++.

mewmew commented:
> So I've applied all the changes agreed on till now and reapplied changes to `.h` files.

@Predelnik That's great! Looks really good. I'll try to review the last half of the `ItemAffixData` on Sunday. Should be ready for merge after then :) Cheers /u

mewmew commented:
> Not sure that is_cursed is PLOk. Seems a bit contradictory.

I had a look and you are right, it should be `ok` rather than `is_cursed`, since `false` => negative effect (i.e. cursed), `true` => positive effect.

mewmew commented:
Do you prefer `is_triggerable` in ItemData, or should we switch to `is_usable`? I slightly lean towards usable since its used in the PSX version.

Still in ItemData, update `price` to `min_price`?

Update `scroll_spell_id` to `spell_id` in ItemData, since it applies both to scrolls and staffs (for staffs it would otherwise be `charge_spell_id`, so `spell_id` is the intersect of names)

mewmew commented:
Ok, I feel done with the review. We can always go back and update things later on when we learn more and gain new insights.

So, these are left to do.

- [ ] `is_cursed` -> `ok` or `positive`
- [ ] `is_triggerable` -> `is_usable`
- [ ] `price` -> `min_price`
- [ ] `scroll_spell_id` -> `spell_id`

Feel free to say if you don't agree with any of these changes, so we can discuss.

Predelnik commented:
> is_cursed -> ok or positive

hmm yeah look like I stupidly overlooked it by copying from freeablo not checking actual values/ instructions which checks it, my bad. However personally I think `ok` and `positive` seems a bit vague, I would even propose `not_cursed`  even if negation in names is usually bad. Maybe `sellable` but this effect also has the other meaning for random generation as far as I remember so I'm not sure.

> is_triggerable -> is_usable

I just forgot to change that so that's totally fine.

> price -> min_price

Hmm I thought this price is the only price that's used so naming it `min` wouldn't really help in understanding what it means for a reader or am I missing something? At least Jarulf seems to list only one price in his tables.

> scroll_spell_id -> spell_id

Hmm I guess I'm ok with it even if the only basic staff having spell is notorious `Short Staff of Charged Bolt`, but I guess this could be put in a comment.

mewmew commented:
> hmm yeah look like I stupidly overlooked it by copying from freeablo not checking actual values/ instructions which checks it, my bad. However personally I think ok and positive seems a bit vague, I would even propose not_cursed even if negation in names is usually bad. Maybe sellable but this effect also has the other meaning for random generation as far as I remember so I'm not sure.

Lets go with `not_cursed` for now.

> I just forgot to change that so that's totally fine.

Good.

> Hmm I thought this price is the only price that's used so naming it `min` wouldn't really help in understanding what it means for a reader or am I missing something? At least Jarulf seems to list only one price in his tables.

Actually, you seem to be right. I confused the `min_cost` and `max_cost` from ItemAffixData with the `price` and `max_price` from ItemData.

The PSX debug info specifies `PLMinVal` and `PLMaxVal` for `min_price` and `max_price` on ItemAffixData, but uses `iValue` and `iMaxValue` for `price` and `max_price` on ItemData.

```c
typedef struct PLStruct {
...
	int PLMinVal;
	int PLMaxVal;
}

typedef struct ItemDataStruct {
...
	unsigned short iValue;
	unsigned short iMaxValue;
}
```

Perhaps we should decide to use either cost or price? But I'm fine with keeping `price` and `max_price` on ItemData, as that is consistent with the PSX debug info.

> Hmm I guess I'm ok with it even if the only basic staff having spell is notorious Short Staff of Charged Bolt, but I guess this could be put in a comment.

Sure, lets change to `spell_id` and put `// Used for charges on Short Staff of Charged Bolt, and used for scrolls otherwise.` in a comment.

Predelnik commented:
So I've applied all these changes and changed `min/max_cost`  to `min/max_price` in `ItemAffixData`. 

Thinking about it - both price and cost mostly similar but good thing about price is that it usually applies to money and cost could be used for other things, for example mana cost. However in case of affixes cost actually may make sense because it's not price of something whole but sort of additional cost provided for affix applied, but I don't think we need to go into this kind of rabbit hole, so I changed it to price in both cases :)

And I haven't commented about it but underscore for `item_class` field is really unfortunate maybe it'd be good to change it somehow in the future.

Yeah hope we finally came to a close on this one. Thanks for all this thorough reviewing :)

mewmew commented:
> So I've applied all these changes and changed min/max_cost to min/max_price in ItemAffixData.
>
> Thinking about it - both price and cost mostly similar but good thing about price is that it usually applies to money and cost could be used for other things, for example mana cost. However in case of affixes cost actually may make sense because it's not price of something whole but sort of additional cost provided for affix applied, but I don't think we need to go into this kind of rabbit hole, so I changed it to price in both cases :)

Sounds good to me :) Thanks for detailing your thoughts.

> And I haven't commented about it but underscore for item_class field is really unfortunate maybe it'd be good to change it somehow in the future.

Yeah, lets keep it for now and look for better alternatives in the future.

> Yeah hope we finally came to a close on this one. Thanks for all this thorough reviewing :)

Likewise, I think we ended up with more knowledge of how these structs actually work based on our discussions. Thanks for being patient :)

I'll merge!
